File: sdk/python/tavok/agent.py
"""Tavok SDK Agent — the main user-facing class.

An Agent registers with a Tavok server, connects via WebSocket, and
responds to messages with optional token streaming.

Minimal example::

    from tavok import Agent

    agent = Agent(name="my-agent")

    @agent.on_mention
    async def handle(msg):
        async with agent.stream(msg.channel_id) as s:
            await s.token("Hello! ")
            await s.token("I'm an agent.")

    agent.run()  # auto-discovers server from .tavok.json or env vars
"""

from __future__ import annotations

import asyncio
import logging
import os
import signal
import sys
from collections.abc import Callable, Coroutine
from typing import Any

from .auth import register_agent
from .config import TavokConfig
from .stream import StreamContext
from .types import Message, StreamComplete, StreamError, StreamStart, StreamToken
from .ws import PhoenixSocket

logger = logging.getLogger("tavok")

# Type alias for event handlers
MessageHandler = Callable[[Message], Coroutine[Any, Any, None]]


class Agent:
    """A Tavok agent that connects via WebSocket and responds to messages.

    All connection parameters are optional. The agent discovers configuration
    from (in order): explicit arguments, environment variables, ``.tavok.json``
    file, then localhost defaults.

    Args:
        url: Gateway WebSocket URL (e.g. ``ws://localhost:4001``).
        api_url: Web server URL for REST API (e.g. ``http://localhost:5555``).
        name: Display name for the agent.
        api_key: Existing API key. If not provided, the agent will register
            on :meth:`start` and receive a new key. Also reads ``TAVOK_API_KEY``.
        agent_id: Existing agent/bot ULID. Required if ``api_key`` is provided.
            Also reads ``TAVOK_AGENT_ID``.
        server_id: Default server ULID. Also discovered from ``.tavok.json``.
        channel_ids: Default channel ULIDs. Also discovered from ``.tavok.json``.
        model: LLM model identifier for display purposes.
        capabilities: List of capability strings.
        avatar_url: Avatar image URL.
    """

    def __init__(
        self,
        *,
        url: str | None = None,
        api_url: str | None = None,
        name: str = "Tavok Agent",
        api_key: str | None = None,
        agent_id: str | None = None,
        server_id: str | None = None,
        channel_ids: list[str] | None = None,
        model: str | None = None,
        capabilities: list[str] | None = None,
        avatar_url: str | None = None,
    ) -> None:
        # Auto-discover configuration from env vars and .tavok.json
        config = TavokConfig.discover()

        self._gateway_url = (url or config.gateway_url).rstrip("/")
        self._api_url = (api_url or config.url).rstrip("/")
        self._name = name
        self._api_key = api_key or os.environ.get("TAVOK_API_KEY")
        self._agent_id = agent_id or os.environ.get("TAVOK_AGENT_ID")
        self._default_server_id = server_id or config.server_id
        self._default_channel_ids = channel_ids or (
            [config.channel_id] if config.channel_id else None
        )
        self._model = model
        self._capabilities = capabilities or ["chat"]
        self._avatar_url = avatar_url

        self._socket: PhoenixSocket | None = None
        self._joined_channels: set[str] = set()
        self._sequences: dict[str, str] = {}  # channel_id -> last sequence

        # Event handlers
        self._on_mention_handlers: list[MessageHandler] = []
        self._on_message_handlers: list[MessageHandler] = []
        self._on_stream_start_handlers: list[Callable] = []
        self._on_stream_token_handlers: list[Callable] = []
        self._on_stream_complete_handlers: list[Callable] = []
        self._on_stream_error_handlers: list[Callable] = []

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def agent_id(self) -> str | None:
        """The agent's bot ULID (available after registration)."""
        return self._agent_id

    @property
    def api_key(self) -> str | None:
        """The agent's API key (available after registration)."""
        return self._api_key

    @property
    def name(self) -> str:
        """The agent's display name."""
        return self._name

    @property
    def connected(self) -> bool:
        """Whether the WebSocket connection is active."""
        return self._socket is not None and self._socket.connected

    # ------------------------------------------------------------------
    # Decorators
    # ------------------------------------------------------------------

    def on_mention(self, fn: MessageHandler) -> MessageHandler:
        """Decorator: register a handler called when the agent is @mentioned.

        Usage::

            @agent.on_mention
            async def handle(msg: Message):
                await agent.send(msg.channel_id, "You mentioned me!")
        """
        self._on_mention_handlers.append(fn)
        return fn

    def on_message(self, fn: MessageHandler) -> MessageHandler:
        """Decorator: register a handler called for every message.

        Usage::

            @agent.on_message
            async def handle(msg: Message):
                if "hello" in msg.content.lower():
                    await agent.send(msg.channel_id, "Hi there!")
        """
        self._on_message_handlers.append(fn)
        return fn

    def on_stream_start(self, fn: Callable) -> Callable:
        """Decorator: called when any bot starts streaming."""
        self._on_stream_start_handlers.append(fn)
        return fn

    def on_stream_complete(self, fn: Callable) -> Callable:
        """Decorator: called when any stream completes."""
        self._on_stream_complete_handlers.append(fn)
        return fn

    # ------------------------------------------------------------------
    # Actions
    # ------------------------------------------------------------------

    async def send(self, channel_id: str, content: str) -> dict:
        """Send a standard (non-streaming) message to a channel.

        Args:
            channel_id: The target channel ULID.
            content: Message text.

        Returns:
            Reply payload with ``id`` and ``sequence``.
        """
        if not self._socket:
            raise RuntimeError("Agent not connected")

        topic = f"room:{channel_id}"
        reply = await self._socket.push(
            topic, "new_message", {"content": content}
        )
        response = reply.get("response", {})
        seq = response.get("sequence")
        if seq:
            self._sequences[channel_id] = str(seq)
        return response

    def stream(
        self, channel_id: str, *, reply_to: str | None = None
    ) -> StreamContext:
        """Create a streaming context for sending tokens word-by-word.

        Usage::

            async with agent.stream(channel_id) as s:
                await s.token("Hello ")
                await s.token("world!")

        Args:
            channel_id: The target channel ULID.
            reply_to: Optional message ID to reply to.

        Returns:
            :class:`StreamContext` async context manager.
        """
        if not self._socket:
            raise RuntimeError("Agent not connected")

        return StreamContext(
            socket=self._socket,
            channel_id=channel_id,
            bot_id=self._agent_id or "",
            bot_name=self._name,
            reply_to=reply_to,
        )

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def start(
        self,
        *,
        server_id: str | None = None,
        channel_ids: list[str] | None = None,
    ) -> None:
        """Register (if needed), connect, and join channels.

        Args:
            server_id: The server ULID to register with. If not provided,
                uses the value from the constructor, env vars, or .tavok.json.
            channel_ids: Channel ULIDs to join. If not provided, uses
                auto-discovered defaults. If empty list, no channels joined.
        """
        resolved_server = server_id or self._default_server_id
        resolved_channels = channel_ids if channel_ids is not None else self._default_channel_ids

        # Step 1: Register if we don't have an API key
        if not self._api_key:
            if not resolved_server:
                raise ValueError(
                    "server_id is required for agent registration. "
                    "Provide it to start(), Agent(), set TAVOK_SERVER_ID, "
                    "or run 'npx tavok init' to create .tavok.json."
                )
            logger.info("Registering agent '%s' with server %s", self._name, resolved_server)
            result = await register_agent(
                base_url=self._api_url,
                server_id=resolved_server,
                display_name=self._name,
                model=self._model,
                capabilities=self._capabilities,
                avatar_url=self._avatar_url,
            )
            self._api_key = result.api_key
            self._agent_id = result.agent_id
            logger.info(
                "Registered: agent_id=%s, api_key=%s...%s",
                self._agent_id,
                self._api_key[:12],
                self._api_key[-4:],
            )

        # Step 2: Connect WebSocket
        ws_url = f"{self._gateway_url}/socket/websocket"
        self._socket = PhoenixSocket(
            ws_url, params={"api_key": self._api_key}
        )

        # Register internal event handlers before connecting
        self._register_event_handlers()

        await self._socket.connect()

        # Step 3: Join channels
        for ch_id in resolved_channels or []:
            await self.join_channel(ch_id)

    async def join_channel(self, channel_id: str) -> None:
        """Join a channel.

        Args:
            channel_id: The channel ULID to join.
        """
        if not self._socket:
            raise RuntimeError("Agent not connected")

        topic = f"room:{channel_id}"
        last_seq = self._sequences.get(channel_id)
        payload = {}
        if last_seq:
            payload["lastSequence"] = last_seq

        await self._socket.join(topic, payload)
        self._joined_channels.add(channel_id)

    async def leave_channel(self, channel_id: str) -> None:
        """Leave a channel."""
        if self._socket:
            await self._socket.leave(f"room:{channel_id}")
        self._joined_channels.discard(channel_id)

    async def stop(self) -> None:
        """Disconnect and clean up."""
        if self._socket:
            await self._socket.disconnect()
            self._socket = None
        self._joined_channels.clear()
        logger.info("Agent stopped")

    def run(
        self,
        *,
        server_id: str | None = None,
        channel_ids: list[str] | None = None,
    ) -> None:
        """Blocking entry point — registers, connects, and runs forever.

        This is the simplest way to run an agent::

            agent = Agent(name="my-agent")
            agent.run()  # auto-discovers from .tavok.json

        Args:
            server_id: The server ULID. Optional if discoverable from
                constructor, env vars, or .tavok.json.
            channel_ids: Channel ULIDs to join.
        """
        async def _main() -> None:
            await self.start(server_id=server_id, channel_ids=channel_ids)
            # Print to stderr so users always see this, even without logging
            print(
                f"Agent '{self._name}' running (id={self._agent_id}, "
                f"gateway={self._gateway_url}). Press Ctrl+C to stop.",
                file=sys.stderr,
            )
            logger.info(
                "Agent '%s' running (id=%s). Press Ctrl+C to stop.",
                self._name,
                self._agent_id,
            )

            # Wait until closed
            stop_event = asyncio.Event()

            def _signal_handler() -> None:
                logger.info("Shutting down...")
                stop_event.set()

            loop = asyncio.get_running_loop()
            for sig in (signal.SIGINT, signal.SIGTERM):
                try:
                    loop.add_signal_handler(sig, _signal_handler)
                except NotImplementedError:
                    # Windows doesn't support add_signal_handler
                    pass

            try:
                await stop_event.wait()
            except KeyboardInterrupt:
                pass
            finally:
                await self.stop()

        try:
            asyncio.run(_main())
        except KeyboardInterrupt:
            pass

    # ------------------------------------------------------------------
    # Internal event routing
    # ------------------------------------------------------------------

    def _register_event_handlers(self) -> None:
        """Wire up Phoenix Channel events to our handler methods."""
        assert self._socket is not None

        self._socket.on("message_new", self._handle_message_new)
        self._socket.on("stream_start", self._handle_stream_start)
        self._socket.on("stream_token", self._handle_stream_token)
        self._socket.on("stream_complete", self._handle_stream_complete)
        self._socket.on("stream_error", self._handle_stream_error)

    async def _handle_message_new(self, topic: str, payload: dict) -> None:
        """Handle incoming message_new broadcasts."""
        msg = Message.from_payload(payload)

        # Don't react to our own messages
        if msg.author_id == self._agent_id:
            return

        # Track sequence
        channel_id = topic.removeprefix("room:")
        if msg.sequence:
            self._sequences[channel_id] = msg.sequence

        # Dispatch to on_message handlers
        for handler in self._on_message_handlers:
            try:
                await handler(msg)
            except Exception:
                logger.exception("on_message handler error")

        # Dispatch to on_mention handlers if mentioned
        if self._agent_id and msg.mentions_me(self._agent_id):
            for handler in self._on_mention_handlers:
                try:
                    await handler(msg)
                except Exception:
                    logger.exception("on_mention handler error")

    async def _handle_stream_start(self, topic: str, payload: dict) -> None:
        """Handle stream_start broadcasts."""
        start = StreamStart.from_payload(payload)
        for handler in self._on_stream_start_handlers:
            try:
                result = handler(start)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                logger.exception("on_stream_start handler error")

    async def _handle_stream_token(self, topic: str, payload: dict) -> None:
        """Handle stream_token broadcasts."""
        token = StreamToken.from_payload(payload)
        for handler in self._on_stream_token_handlers:
            try:
                result = handler(token)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                logger.exception("on_stream_token handler error")

    async def _handle_stream_complete(self, topic: str, payload: dict) -> None:
        """Handle stream_complete broadcasts."""
        complete = StreamComplete.from_payload(payload)
        for handler in self._on_stream_complete_handlers:
            try:
                result = handler(complete)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                logger.exception("on_stream_complete handler error")

    async def _handle_stream_error(self, topic: str, payload: dict) -> None:
        """Handle stream_error broadcasts."""
        error = StreamError.from_payload(payload)
        for handler in self._on_stream_error_handlers:
            try:
                result = handler(error)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                logger.exception("on_stream_error handler error")

File: packages/sdk/src/agent.ts
/**
 * Tavok SDK Agent — the main user-facing class.
 *
 * An Agent connects via WebSocket and responds to messages with optional
 * token streaming. Credentials are auto-discovered from `.tavok-agents.json`
 * (written by `tavok init`) or provided via environment variables.
 *
 * Ported from the Python SDK (`sdk/python/tavok/agent.py`).
 *
 * @example
 * ```typescript
 * import { Agent } from "@tavok/sdk";
 *
 * const agent = new Agent({ name: "Jack" });
 *
 * agent.onMention(async (msg) => {
 *   const ctx = agent.stream(msg.channelId);
 *   await ctx.start();
 *   await ctx.token("Hello! I'm an agent.");
 *   await ctx.finish();
 * });
 *
 * agent.run();
 * ```
 */

import { PhoenixSocket } from "./phoenix";
import { StreamContext } from "./stream";
import { TavokConfig } from "./config";
import { discoverCredentials } from "./auth";
import {
  type Message,
  type StreamStart,
  type StreamComplete,
  type StreamError,
  messageFromPayload,
  streamStartFromPayload,
  streamCompleteFromPayload,
  streamErrorFromPayload,
} from "./types";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Options for creating an Agent. */
export interface AgentOptions {
  /** Agent display name. Used for auto-discovery from `.tavok-agents.json`. */
  name?: string;
  /** Gateway WebSocket URL (e.g. `ws://localhost:4001`). */
  url?: string;
  /** Web server URL for REST API (e.g. `http://localhost:5555`). */
  apiUrl?: string;
  /** API key. If not provided, auto-discovered by name. */
  apiKey?: string;
  /** Agent ULID. If not provided, auto-discovered by name. */
  agentId?: string;
  /** Default server ULID. */
  serverId?: string;
  /** Default channel ULIDs to join. */
  channelIds?: string[];
}

/** Handler called with a parsed Message. */
export type MessageHandler = (msg: Message) => void | Promise<void>;

/** Handler called when any agent starts streaming. */
export type StreamStartHandler = (start: StreamStart) => void | Promise<void>;

/** Handler called when any stream completes. */
export type StreamCompleteHandler = (complete: StreamComplete) => void | Promise<void>;

/** Handler called when a stream errors. */
export type StreamErrorHandler = (error: StreamError) => void | Promise<void>;

// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------

/**
 * A Tavok agent that connects via WebSocket and responds to messages.
 *
 * Credentials are resolved in order:
 *
 * 1. Explicit constructor arguments (`apiKey`, `agentId`)
 * 2. Environment variables (`TAVOK_API_KEY`, `TAVOK_AGENT_ID`)
 * 3. Auto-discovery from `.tavok-agents.json` (by `name`)
 * 4. `TavokConfig.discover()` for topology
 */
export class Agent {
  private readonly _gatewayUrl: string;
  private readonly _apiUrl: string;
  private readonly _name: string;
  private _apiKey: string | undefined;
  private _agentId: string | undefined;
  private readonly _defaultServerId: string | undefined;
  private readonly _defaultChannelIds: string[] | undefined;

  private _socket: PhoenixSocket | null = null;
  private readonly _joinedChannels = new Set<string>();
  private readonly _sequences = new Map<string, string>();

  // Handler lists
  private readonly _onMentionHandlers: MessageHandler[] = [];
  private readonly _onMessageHandlers: MessageHandler[] = [];
  private readonly _onStreamStartHandlers: StreamStartHandler[] = [];
  private readonly _onStreamCompleteHandlers: StreamCompleteHandler[] = [];
  private readonly _onStreamErrorHandlers: StreamErrorHandler[] = [];

  // ------------------------------------------------------------------
  // Constructor
  // ------------------------------------------------------------------

  constructor(opts?: AgentOptions) {
    const o = opts ?? {};

    // Auto-discover topology from env vars and .tavok.json
    const config = TavokConfig.discover();

    this._gatewayUrl = (o.url ?? config.gatewayUrl).replace(/\/+$/, "");
    this._apiUrl = (o.apiUrl ?? config.url).replace(/\/+$/, "");
    this._name = o.name ?? "Tavok Agent";

    // Credential resolution: explicit -> env -> file discovery
    this._apiKey = o.apiKey ?? process.env.TAVOK_API_KEY;
    this._agentId = o.agentId ?? process.env.TAVOK_AGENT_ID;

    if (!this._apiKey) {
      const creds = discoverCredentials(this._name);
      if (creds) {
        this._apiKey = creds.apiKey || undefined;
        this._agentId = this._agentId ?? (creds.id || undefined);
      }
    }

    this._defaultServerId = o.serverId ?? config.serverId ?? undefined;
    this._defaultChannelIds =
      o.channelIds ?? (config.channelId ? [config.channelId] : undefined);
  }

  // ------------------------------------------------------------------
  // Properties
  // ------------------------------------------------------------------

  /** The agent's ULID. */
  get agentId(): string | undefined {
    return this._agentId;
  }

  /** The agent's API key. */
  get apiKey(): string | undefined {
    return this._apiKey;
  }

  /** The agent's display name. */
  get name(): string {
    return this._name;
  }

  /** Whether the WebSocket connection is active. */
  get connected(): boolean {
    return this._socket !== null && this._socket.connected;
  }

  // ------------------------------------------------------------------
  // Handler registration (fluent / chainable)
  // ------------------------------------------------------------------

  /**
   * Register a handler called when the agent is @mentioned.
   *
   * Mention is detected by checking if the message content contains
   * `<@{agentId}>`.
   */
  onMention(handler: MessageHandler): this {
    this._onMentionHandlers.push(handler);
    return this;
  }

  /** Register a handler called for every incoming message. */
  onMessage(handler: MessageHandler): this {
    this._onMessageHandlers.push(handler);
    return this;
  }

  /** Register a handler called when any agent starts streaming. */
  onStreamStart(handler: StreamStartHandler): this {
    this._onStreamStartHandlers.push(handler);
    return this;
  }

  /** Register a handler called when any stream completes. */
  onStreamComplete(handler: StreamCompleteHandler): this {
    this._onStreamCompleteHandlers.push(handler);
    return this;
  }

  /** Register a handler called when a stream errors. */
  onStreamError(handler: StreamErrorHandler): this {
    this._onStreamErrorHandlers.push(handler);
    return this;
  }

  // ------------------------------------------------------------------
  // Actions
  // ------------------------------------------------------------------

  /**
   * Send a standard (non-streaming) message to a channel.
   *
   * @param channelId - The target channel ULID.
   * @param content   - Message text.
   * @returns Reply payload from the server.
   */
  async send(
    channelId: string,
    content: string,
  ): Promise<Record<string, unknown>> {
    if (!this._socket) {
      throw new Error("Agent not connected");
    }

    const topic = `room:${channelId}`;
    const reply = await this._socket.push(topic, "new_message", { content });
    const response = (reply.response as Record<string, unknown>) ?? reply;
    const seq = response.sequence as string | undefined;
    if (seq) {
      this._sequences.set(channelId, String(seq));
    }
    return response;
  }

  /**
   * Create a streaming context for sending tokens word-by-word.
   *
   * The caller is responsible for calling `ctx.start()` to begin the stream.
   *
   * @param channelId - The target channel ULID.
   * @param opts      - Optional `replyTo` message ID.
   * @returns A new StreamContext (not yet started).
   */
  stream(
    channelId: string,
    opts?: { replyTo?: string },
  ): StreamContext {
    if (!this._socket) {
      throw new Error("Agent not connected");
    }

    return new StreamContext(
      this._socket,
      channelId,
      this._agentId ?? "",
      this._name,
      opts?.replyTo,
    );
  }

  // ------------------------------------------------------------------
  // Lifecycle
  // ------------------------------------------------------------------

  /**
   * Connect to the gateway and join channels.
   *
   * @param opts - Optional overrides for serverId and channelIds.
   */
  async start(opts?: {
    serverId?: string;
    channelIds?: string[];
  }): Promise<void> {
    const resolvedChannels =
      opts?.channelIds ?? this._defaultChannelIds;

    if (!this._apiKey) {
      throw new Error(
        `No API key found for agent '${this._name}'. ` +
          "Run 'tavok init' to create agent credentials in " +
          ".tavok-agents.json, set TAVOK_API_KEY env var, " +
          "or pass apiKey to Agent().",
      );
    }

    // Connect WebSocket
    const wsUrl = `${this._gatewayUrl}/socket/websocket`;
    this._socket = new PhoenixSocket(wsUrl, { api_key: this._apiKey });

    // Register internal event handlers before connecting
    this._registerEventHandlers();

    await this._socket.connect();

    // Join channels
    for (const chId of resolvedChannels ?? []) {
      await this.joinChannel(chId);
    }
  }

  /** Join a channel by its ULID. */
  async joinChannel(channelId: string): Promise<void> {
    if (!this._socket) {
      throw new Error("Agent not connected");
    }

    const topic = `room:${channelId}`;
    const payload: Record<string, unknown> = {};
    const lastSeq = this._sequences.get(channelId);
    if (lastSeq) {
      payload.lastSequence = lastSeq;
    }

    await this._socket.join(topic, payload);
    this._joinedChannels.add(channelId);
  }

  /** Leave a channel by its ULID. */
  async leaveChannel(channelId: string): Promise<void> {
    if (this._socket) {
      await this._socket.leave(`room:${channelId}`);
    }
    this._joinedChannels.delete(channelId);
  }

  /** Disconnect and clean up. */
  async stop(): Promise<void> {
    if (this._socket) {
      await this._socket.disconnect();
      this._socket = null;
    }
    this._joinedChannels.clear();
  }

  /**
   * Blocking entry point — connects and runs until SIGINT/SIGTERM.
   *
   * @param opts - Optional overrides for serverId and channelIds.
   */
  run(opts?: { serverId?: string; channelIds?: string[] }): void {
    const execute = async (): Promise<void> => {
      await this.start(opts);

      process.stderr.write(
        `Agent '${this._name}' running (id=${this._agentId}, ` +
          `gateway=${this._gatewayUrl}). Press Ctrl+C to stop.\n`,
      );

      // Wait until signal
      await new Promise<void>((resolve) => {
        const handler = (): void => {
          resolve();
        };

        process.on("SIGINT", handler);
        process.on("SIGTERM", handler);
      });

      await this.stop();
    };

    execute().catch((err: unknown) => {
      process.stderr.write(
        `Agent error: ${err instanceof Error ? err.message : String(err)}\n`,
      );
      process.exitCode = 1;
    });
  }

  // ------------------------------------------------------------------
  // Internal event routing
  // ------------------------------------------------------------------

  /** Wire up Phoenix Channel events to internal handler methods. */
  private _registerEventHandlers(): void {
    if (!this._socket) return;

    this._socket.on(
      "message_new",
      (topic: string, payload: Record<string, unknown>) => {
        this._handleMessageNew(topic, payload);
      },
    );
    this._socket.on(
      "stream_start",
      (_topic: string, payload: Record<string, unknown>) => {
        this._handleStreamStart(payload);
      },
    );
    this._socket.on(
      "stream_complete",
      (_topic: string, payload: Record<string, unknown>) => {
        this._handleStreamComplete(payload);
      },
    );
    this._socket.on(
      "stream_error",
      (_topic: string, payload: Record<string, unknown>) => {
        this._handleStreamError(payload);
      },
    );
  }

  /** Handle incoming message_new broadcasts. */
  private async _handleMessageNew(
    topic: string,
    payload: Record<string, unknown>,
  ): Promise<void> {
    const msg = messageFromPayload(payload);

    // Don't react to our own messages
    if (msg.authorId === this._agentId) {
      return;
    }

    // Track sequence
    const channelId = topic.replace(/^room:/, "");
    if (msg.sequence) {
      this._sequences.set(channelId, msg.sequence);
    }

    // Dispatch to onMessage handlers
    for (const handler of this._onMessageHandlers) {
      try {
        await handler(msg);
      } catch (err) {
        // Handler exceptions are caught and logged — don't crash
        console.error("onMessage handler error:", err);
      }
    }

    // Dispatch to onMention handlers if mentioned
    if (this._agentId && msg.content.includes(`<@${this._agentId}>`)) {
      for (const handler of this._onMentionHandlers) {
        try {
          await handler(msg);
        } catch (err) {
          console.error("onMention handler error:", err);
        }
      }
    }
  }

  /** Handle stream_start broadcasts. */
  private async _handleStreamStart(
    payload: Record<string, unknown>,
  ): Promise<void> {
    const start = streamStartFromPayload(payload);
    for (const handler of this._onStreamStartHandlers) {
      try {
        await handler(start);
      } catch (err) {
        console.error("onStreamStart handler error:", err);
      }
    }
  }

  /** Handle stream_complete broadcasts. */
  private async _handleStreamComplete(
    payload: Record<string, unknown>,
  ): Promise<void> {
    const complete = streamCompleteFromPayload(payload);
    for (const handler of this._onStreamCompleteHandlers) {
      try {
        await handler(complete);
      } catch (err) {
        console.error("onStreamComplete handler error:", err);
      }
    }
  }

  /** Handle stream_error broadcasts. */
  private async _handleStreamError(
    payload: Record<string, unknown>,
  ): Promise<void> {
    const error = streamErrorFromPayload(payload);
    for (const handler of this._onStreamErrorHandlers) {
      try {
        await handler(error);
      } catch (err) {
        console.error("onStreamError handler error:", err);
      }
    }
  }
}
